File: 0988-smallest-string-starting-from-leaf/0988-smallest-string-starting-from-leaf.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def solve(self, root, curr, ans):
        if not root:
            return
        curr = chr(ord('a') + root.val) + curr
        if not root.left and not root.right:
            if ans[0] == "":
                ans[0] = curr
            elif ans[0] > curr:
                ans[0] = curr

        self.solve(root.left, curr, ans)
        self.solve(root.right, curr, ans)


    def smallestFromLeaf(self, root):
        if not root:
            return ""
        ans = [""]
        self.solve(root, "", ans)
        return ans[0]
        

File: 0988-smallest-string-starting-from-leaf/0988-smallest-string-starting-from-leaf.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void solve(TreeNode* root, string curr, string &ans) {
        if(!root) return;
        curr = (char)('a' + root->val) + curr;
        if(!root->left && !root->right) {
           if(ans == "") ans = curr;
           else if(ans > curr) ans = curr;
        }

        solve(root->left, curr, ans);
        solve(root->right, curr, ans);
    }

    string smallestFromLeaf(TreeNode* root) {
        if(!root) return "";
        string ans = "";
        solve(root, "", ans);
        return ans;
    }
};